File: lattice_zk/poly.py
"""
Polynomial Arithmetic over R_q = Z_q[X]/(X^n + 1).

Implements coefficient-level operations for the lattice-native ZKPoK.
Uses schoolbook multiplication (O(n²)) for the research prototype.
A production implementation would use NTT-based O(n log n) multiplication.

All polynomials are represented as lists of n integers in [0, q).
Module vectors are lists of polynomials.
Module matrices are lists of lists of polynomials (row-major).
"""

from typing import List
from .params import LatticeZKParams, DEFAULT_ZK_PARAMS

# Type aliases for clarity
Poly = List[int]            # Single polynomial: n coefficients in [0, q)
PolyVec = List[Poly]        # Vector of polynomials (length k or l)
PolyMat = List[PolyVec]     # Matrix of polynomials (k × l)


# ═══════════════════════════════════════════════════════════════
# Scalar / Coefficient Operations
# ═══════════════════════════════════════════════════════════════

def mod_centered(x: int, q: int) -> int:
    """Map x ∈ Z_q to centered representation in [-(q-1)/2, (q-1)/2]."""
    r = x % q
    if r > q // 2:
        r -= q
    return r


def coeff_norm_inf(coeffs: List[int], q: int) -> int:
    """Compute ||coeffs||∞ in centered representation."""
    return max(abs(mod_centered(c, q)) for c in coeffs)


# ═══════════════════════════════════════════════════════════════
# Polynomial Operations in R_q = Z_q[X]/(X^n + 1)
# ═══════════════════════════════════════════════════════════════

def poly_zero(n: int) -> Poly:
    """Return the zero polynomial."""
    return [0] * n


def poly_add(a: Poly, b: Poly, q: int) -> Poly:
    """Polynomial addition: (a + b) mod q."""
    return [(a[i] + b[i]) % q for i in range(len(a))]


def poly_sub(a: Poly, b: Poly, q: int) -> Poly:
    """Polynomial subtraction: (a - b) mod q."""
    return [(a[i] - b[i]) % q for i in range(len(a))]


def poly_neg(a: Poly, q: int) -> Poly:
    """Polynomial negation: (-a) mod q."""
    return [(-c) % q for c in a]


def poly_mul(a: Poly, b: Poly, q: int, n: int) -> Poly:
    """
    Polynomial multiplication in R_q = Z_q[X]/(X^n + 1).

    Schoolbook algorithm with reduction mod (X^n + 1).
    Uses the identity: X^n ≡ -1 (mod X^n + 1).

    Complexity: O(n²). Production code would use NTT.
    """
    c = [0] * n
    for i in range(n):
        if a[i] == 0:
            continue
        for j in range(n):
            if b[j] == 0:
                continue
            idx = i + j
            prod = a[i] * b[j]
            if idx < n:
                c[idx] = (c[idx] + prod) % q
            else:
                # X^(n+r) = X^r · X^n ≡ -X^r (mod X^n + 1)
                c[idx - n] = (c[idx - n] - prod) % q
    return c


def poly_scale(a: Poly, scalar: int, q: int) -> Poly:
    """Scalar multiplication: scalar · a mod q."""
    return [(scalar * c) % q for c in a]


def poly_norm_inf(a: Poly, q: int) -> int:
    """Infinity norm of polynomial in centered representation."""
    return coeff_norm_inf(a, q)


# ═══════════════════════════════════════════════════════════════
# Module Vector Operations
# ═══════════════════════════════════════════════════════════════

def polyvec_add(a: PolyVec, b: PolyVec, q: int) -> PolyVec:
    """Component-wise vector addition."""
    return [poly_add(a[i], b[i], q) for i in range(len(a))]


def polyvec_sub(a: PolyVec, b: PolyVec, q: int) -> PolyVec:
    """Component-wise vector subtraction."""
    return [poly_sub(a[i], b[i], q) for i in range(len(a))]


def polyvec_neg(a: PolyVec, q: int) -> PolyVec:
    """Component-wise negation."""
    return [poly_neg(p, q) for p in a]


def polyvec_norm_inf(v: PolyVec, q: int) -> int:
    """Infinity norm of a polynomial vector (max over all coefficients)."""
    return max(poly_norm_inf(p, q) for p in v)


# ═══════════════════════════════════════════════════════════════
# Matrix-Vector Multiplication
# ═══════════════════════════════════════════════════════════════

def matvec_mul(A: PolyMat, v: PolyVec, q: int, n: int) -> PolyVec:
    """
    Matrix-vector multiplication: w = A · v in R_q^k.

    A is k×l, v is length l, result is length k.
    Each entry is a polynomial in R_q.
    """
    k = len(A)
    l = len(A[0])
    assert len(v) == l, f"Dimension mismatch: A is {k}×{l}, v is {len(v)}"

    result = [poly_zero(n) for _ in range(k)]
    for i in range(k):
        for j in range(l):
            prod = poly_mul(A[i][j], v[j], q, n)
            result[i] = poly_add(result[i], prod, q)
    return result


def polyvec_inner(a: PolyVec, b: PolyVec, q: int, n: int) -> Poly:
    """Inner product of two polynomial vectors: <a, b> = Σ a_i · b_i."""
    assert len(a) == len(b)
    result = poly_zero(n)
    for i in range(len(a)):
        prod = poly_mul(a[i], b[i], q, n)
        result = poly_add(result, prod, q)
    return result


# ═══════════════════════════════════════════════════════════════
# Serialization (for hashing / proof encoding)
# ═══════════════════════════════════════════════════════════════

def poly_to_bytes(p: Poly, q: int) -> bytes:
    """
    Serialize polynomial coefficients to bytes.

    Each coefficient is encoded as 3 bytes (24 bits) in little-endian,
    sufficient for q = 8380417 < 2^23.
    """
    buf = bytearray()
    for c in p:
        val = c % q  # ensure positive
        buf.append(val & 0xFF)
        buf.append((val >> 8) & 0xFF)
        buf.append((val >> 16) & 0xFF)
    return bytes(buf)


def poly_from_bytes(data: bytes, n: int, q: int) -> Poly:
    """Deserialize polynomial from bytes (3 bytes per coefficient)."""
    assert len(data) == n * 3
    p = []
    for i in range(n):
        val = data[3*i] | (data[3*i+1] << 8) | (data[3*i+2] << 16)
        p.append(val % q)
    return p


def polyvec_to_bytes(v: PolyVec, q: int) -> bytes:
    """Serialize a polynomial vector."""
    return b"".join(poly_to_bytes(p, q) for p in v)


def polyvec_from_bytes(data: bytes, dim: int, n: int, q: int) -> PolyVec:
    """Deserialize a polynomial vector."""
    poly_size = n * 3
    assert len(data) == dim * poly_size
    return [poly_from_bytes(data[i*poly_size:(i+1)*poly_size], n, q)
            for i in range(dim)]


def encode_z(z: PolyVec, gamma1: int, n: int, q: int) -> bytes:
    """
    Encode response vector z with coefficients in [-γ₁+1, γ₁].
    """
    buf = bytearray()
    for poly in z:
        for coeff in poly:
            # First map to centered representation (if it was [0, q))
            c_centered = mod_centered(coeff, q)
            # Shift to positive: val = γ₁ - c_centered
            val = gamma1 - c_centered
            buf.extend(val.to_bytes(3, 'little', signed=False))
    return bytes(buf)


def decode_z(data: bytes, dim: int, n: int, gamma1: int) -> PolyVec:
    """Decode response vector z from bytes."""
    z = []
    offset = 0
    for _ in range(dim):
        poly = []
        for _ in range(n):
            val = int.from_bytes(data[offset:offset+3], 'little')
            coeff = gamma1 - val
            poly.append(coeff)
            offset += 3
        z.append(poly)
    return z

File: eirn-kcp-rs/src/poly.rs
//! Polynomial Types and NTT Stubs for R_q = Z_q[X]/(X^256 + 1).
//!
//! Provides the foundational types for lattice arithmetic.
//! NTT implementation is stubbed — a production build would use
//! optimized NTT from the Dilithium reference implementation.

use crate::params::{N, Q};

// ═══════════════════════════════════════════════════════════════
// Core Polynomial Type
// ═══════════════════════════════════════════════════════════════

/// A polynomial in R_q = Z_q[X]/(X^256 + 1).
///
/// Coefficients stored in standard (non-NTT) representation,
/// each in [0, Q). Centered representation uses [-Q/2, Q/2].
#[derive(Clone, Debug)]
pub struct Poly {
    pub coeffs: [i32; N],
}

impl Poly {
    /// Zero polynomial.
    pub const fn zero() -> Self {
        Poly { coeffs: [0i32; N] }
    }

    /// Reduce all coefficients mod Q to [0, Q).
    pub fn reduce(&mut self) {
        for c in self.coeffs.iter_mut() {
            *c = ((*c % Q as i32) + Q as i32) as i32 % Q as i32;
        }
    }

    /// Infinity norm in centered representation.
    pub fn norm_inf(&self) -> u32 {
        self.coeffs.iter()
            .map(|&c| {
                let r = c.rem_euclid(Q as i32);
                if r > Q as i32 / 2 { Q as i32 - r } else { r }
            })
            .max()
            .unwrap_or(0) as u32
    }

    /// Polynomial addition: (self + other) mod Q.
    pub fn add(&self, other: &Poly) -> Poly {
        let mut result = Poly::zero();
        for i in 0..N {
            result.coeffs[i] = (self.coeffs[i] + other.coeffs[i]).rem_euclid(Q as i32);
        }
        result
    }

    /// Polynomial subtraction: (self - other) mod Q.
    pub fn sub(&self, other: &Poly) -> Poly {
        let mut result = Poly::zero();
        for i in 0..N {
            result.coeffs[i] = (self.coeffs[i] - other.coeffs[i]).rem_euclid(Q as i32);
        }
        result
    }

    /// Schoolbook multiplication in R_q = Z_q[X]/(X^N + 1).
    ///
    /// O(N²) — production code should use NTT-based multiplication.
    pub fn mul(&self, other: &Poly) -> Poly {
        let mut result = Poly::zero();
        for i in 0..N {
            if self.coeffs[i] == 0 { continue; }
            for j in 0..N {
                if other.coeffs[j] == 0 { continue; }
                let prod = (self.coeffs[i] as i64) * (other.coeffs[j] as i64);
                let idx = i + j;
                if idx < N {
                    result.coeffs[idx] = ((result.coeffs[idx] as i64 + prod) % Q as i64) as i32;
                } else {
                    // X^N ≡ -1 mod (X^N + 1)
                    result.coeffs[idx - N] = ((result.coeffs[idx - N] as i64 - prod) % Q as i64) as i32;
                }
            }
        }
        result.reduce();
        result
    }

    /// NTT forward transform (STUB — to be implemented).
    pub fn ntt(&mut self) {
        // TODO: Implement NTT for q = 8380417
        // Uses primitive 512th root of unity: ζ = 1753
        unimplemented!("NTT forward transform")
    }

    /// NTT inverse transform (STUB — to be implemented).
    pub fn inv_ntt(&mut self) {
        unimplemented!("NTT inverse transform")
    }

    /// Serialize to bytes (3 bytes per coefficient, little-endian).
    pub fn to_bytes(&self) -> [u8; N * 3] {
        let mut buf = [0u8; N * 3];
        for i in 0..N {
            let val = self.coeffs[i].rem_euclid(Q as i32) as u32;
            buf[3 * i] = (val & 0xFF) as u8;
            buf[3 * i + 1] = ((val >> 8) & 0xFF) as u8;
            buf[3 * i + 2] = ((val >> 16) & 0xFF) as u8;
        }
        buf
    }

    /// Deserialize from bytes.
    pub fn from_bytes(data: &[u8; N * 3]) -> Self {
        let mut p = Poly::zero();
        for i in 0..N {
            let val = data[3 * i] as u32
                | ((data[3 * i + 1] as u32) << 8)
                | ((data[3 * i + 2] as u32) << 16);
            p.coeffs[i] = (val % Q) as i32;
        }
        p
    }
}

// ═══════════════════════════════════════════════════════════════
// Polynomial Vector (Module Element)
// ═══════════════════════════════════════════════════════════════

/// A vector of polynomials in R_q^k.
#[derive(Clone, Debug)]
pub struct PolyVec<const DIM: usize> {
    pub polys: [Poly; DIM],
}

impl<const DIM: usize> PolyVec<DIM> {
    /// Zero vector.
    pub fn zero() -> Self {
        PolyVec {
            polys: core::array::from_fn(|_| Poly::zero()),
        }
    }

    /// Component-wise addition.
    pub fn add(&self, other: &Self) -> Self {
        PolyVec {
            polys: core::array::from_fn(|i| self.polys[i].add(&other.polys[i])),
        }
    }

    /// Component-wise subtraction.
    pub fn sub(&self, other: &Self) -> Self {
        PolyVec {
            polys: core::array::from_fn(|i| self.polys[i].sub(&other.polys[i])),
        }
    }

    /// Infinity norm (max over all polynomial norms).
    pub fn norm_inf(&self) -> u32 {
        self.polys.iter().map(|p| p.norm_inf()).max().unwrap_or(0)
    }
}

// ═══════════════════════════════════════════════════════════════
// Matrix-Vector Multiplication
// ═══════════════════════════════════════════════════════════════

/// Matrix A ∈ R_q^{K×L} represented as K rows of L polynomials.
#[derive(Clone, Debug)]
pub struct PolyMat<const K: usize, const L: usize> {
    pub rows: [PolyVec<L>; K],
}

impl<const K: usize, const L: usize> PolyMat<K, L> {
    /// Matrix-vector multiplication: A · v ∈ R_q^K.
    pub fn mul_vec(&self, v: &PolyVec<L>) -> PolyVec<K> {
        PolyVec {
            polys: core::array::from_fn(|i| {
                let mut acc = Poly::zero();
                for j in 0..L {
                    let prod = self.rows[i].polys[j].mul(&v.polys[j]);
                    acc = acc.add(&prod);
                }
                acc
            }),
        }
    }
}

// ═══════════════════════════════════════════════════════════════
// Rounding (Decompose / HighBits / LowBits)
// ═══════════════════════════════════════════════════════════════

/// Decompose r into (high, low) parts: r = r₁·α + r₀.
pub fn decompose(r: i32, alpha: u32) -> (i32, i32) {
    let r_pos = r.rem_euclid(Q as i32);
    let mut r0 = r_pos % alpha as i32;
    if r0 > alpha as i32 / 2 {
        r0 -= alpha as i32;
    }
    let mut r1 = (r_pos - r0) / alpha as i32;

    let m = (Q - 1) / alpha;
    if r1 == m as i32 {
        r1 = 0;
        r0 -= 1;
    }
    (r1, r0)
}

/// Extract high-order bits.
pub fn high_bits(r: i32, alpha: u32) -> i32 {
    decompose(r, alpha).0
}

/// Extract low-order bits.
pub fn low_bits(r: i32, alpha: u32) -> i32 {
    decompose(r, alpha).1
}

/// MakeHint: returns 1 if HighBits(r) ≠ HighBits(r + z).
pub fn make_hint(z: i32, r: i32, alpha: u32) -> u8 {
    let r1 = high_bits(r, alpha);
    let v1 = high_bits((r + z).rem_euclid(Q as i32), alpha);
    if r1 != v1 { 1 } else { 0 }
}

/// UseHint: adjust HighBits based on hint.
pub fn use_hint(hint: u8, r: i32, alpha: u32) -> i32 {
    let m = ((Q - 1) / alpha) as i32;
    let (r1, r0) = decompose(r, alpha);
    if hint == 0 {
        r1
    } else if r0 > 0 {
        (r1 + 1).rem_euclid(m)
    } else {
        (r1 - 1).rem_euclid(m)
    }
}
